### PR TITLE
Bucket Notification Import

### PR DESCRIPTION
**Pull request by ravindk89**
# Description

This is the first-pass at migrating https://docs.min.io/docs/minio-bucket-notification-guide.html to the new docs format. There are three big changes here:

1) Organization : I'm experimenting with a format that breaks up the contents of the monolithic bucket notification page into more bite-sized, searchable/findable articles. Specifically:

- Place majority of bucket notification information under MONITORING header
- Keep the Server Features section as a high-level overview + quick links

The goal is to quickly get users to either the conceptual or tutorial pages depending on whether they want to *learn* or *do*. 

Please *ignore* the page listed under `Server Features -> Bucket Notifications`. This is the legacy guide and is just there for helping me track the progress of migration.

2) Content : While the majority of the content is unchanged from what already exists, I've tried to include more prescriptive guidance for setting some of these variables.

3) `mc admin config` reference docs are being built out as part of this work, since its convenient to start seeding it here.

# Goals

The final skeleton of structure should be similar to the following:

- Server Features - High Level description of Bucket Notifications, link to new Bucket Notification guide
- Monitoring
  - Bucket Notifications (Concept Page + links to tutorials)
    - Publish Events to AMQP
    - Publish Events to ...
- MinIO Server (Reference information, Environment Variables
- MinIO Client Admin (mc admin)
  - `mc admin config` (Reference information, Configuration Settings) 

# Build Instructions

The MinIO docs require Python 3 and Node for building. For Node, use v14.5.0 or later.

1) Clone the repo, Checkout the branch (`git checkout -B bucket-notification-migration origin/bucket-notification-migration`)
2) Run `npm install` from the branch
3) Create a python venv (`python -m venv venv`)
4) Install requirements (`pip install -r requirements.txt`)
5) Run `make html`
6) Run `python -m http.server --directory build/bucket-notification-migration/html`
7) Open 127.0.0.1:8000 in your browser and take a poke around.

## Key URLs

- http://127.0.0.1:8000/monitoring/bucket-notifications/bucket-notifications.html
- http://127.0.0.1:8000/reference/minio-server/minio-server.html#amqp-bucket-notification-target
- http://127.0.0.1:8000/reference/minio-cli/minio-mc-admin/mc-admin.config.html#mc-conf.notify_amqp
- http://127.0.0.1:8000/monitoring/bucket-notifications/publish-events-to-amqp.html#minio-bucket-notifications-publish-amqp

### CODE REVIEW

**Comment by ravindk89** [comment]
@poornas @vadmeste this is a WIP pull request, but I'd still like a quick eyeball to see whether there are objections to this format. 

Right now I'm stopping at just the AMQP bucket notification tutorial because I don't want to do all of it and then get feedback to scrap/start over :)

TL;DR Please take a look and see if this looks good from at least a technical perspective. If you also feel that the overall new organization is OK, that would be a nice endorsement

File: sphinxext/minio.py [hunk 1]
```diff
@@ -263,6 +263,8 @@ class MinioObject(ObjectDescription):
     option_spec = {
```
Comment (ravindk89) [inline sphinxext/minio.py:263]: Feel free to ignore changes to this page. They are largely to correct or improve the directive/roles for better linking.

**Comment by ravindk89** [comment]
Made a small update. To apply the patch:

1) `git pull --rebase origin bucket-notification-migration`
2) `rm -rf build/bucket-notification-migration
3) `make html`

**Comment by poornas** [comment]
> @poornas @vadmeste this is a WIP pull request, but I'd still like a quick eyeball to see whether there are objections to this format.
> 
> Right now I'm stopping at just the AMQP bucket notification tutorial because I don't want to do all of it and then get feedback to scrap/start over :)
> 
> TL;DR Please take a look and see if this looks good from at least a technical perspective. If you also feel that the overall new organization is OK, that would be a nice endorsement

@ravindk89 , the information flows nicely - it is well organized. Only one feedback on the `http://127.0.0.1:8000/concepts/bucket-notifications.html` path - it is listing only a subset of the notifications. It would be good to mention additional events such as bucket/replication/ilm and link to the Monitoring page.

**Comment by ravindk89** [comment]
@poornas do we doc those additional events anywhere? 

I was able to derive the current list from [this doc](https://github.com/minio/minio/tree/master/docs/bucket/notifications) - is there a better source?

File: source/monitoring/bucket-notifications/publish-events-to-amqp.rst [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+           internal="<string>" \
+           auto_deleted="<string>" \
+           delivery_mode="<string>" \
+           queue_dir="<string>" \
+           queue_limit="<string>" \
+           comment="<string>"
+
+      - Replace ``IDENTIFIER`` with a unique descriptive string for the
+        AMQP service endpoint. The following examples in this procedure
+        assume an identifier of ``primary``.
```
Comment (eco-minio) [inline source/monitoring/bucket-notifications/publish-events-to-amqp.rst:154]:
```suggestion
        assume an identifier of ``PRIMARY``.
```

File: source/monitoring/bucket-notifications/publish-events-to-amqp.rst [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+2) Restart the MinIO Deployment
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You must restart the MinIO deployment to apply the configuration changes. 
+Use the :mc-cmd:`mc admin service restart` command to restart the deployment.
+
+.. important::
+
+   MinIO restarts *all* :mc:`minio server` processes associated to the 
+   deployment at the same time. Applications may experience a brief period of
```
Comment (eco-minio) [inline source/monitoring/bucket-notifications/publish-events-to-amqp.rst:180]: Not sure if we need to say application will experience interruption due to the fact that it is only for a few seconds and applications are built to retry so this should be invisible to the application.

**Comment by poornas** [comment]
> @poornas do we doc those additional events anywhere?
> 
> I was able to derive the current list from [this doc](https://github.com/minio/minio/tree/master/docs/bucket/notifications) - is there a better source?

No, this is the complete list. http://127.0.0.1:8000/concepts/bucket-notifications.html does not show all the events - it seems to be the intent to leave monitoring events such as replication/ilm to the monitoring page. Just suggesting linking to it since the page showing how to configure notifications doesn't list them all. If it is still wip - please feel free to ignore.

**Review by eco-minio — COMMENTED** [review]
Few comments

File: source/monitoring/bucket-notifications/publish-events-to-amqp.rst (hunk 3 above)
Comment (ravindk89) [reply source/monitoring/bucket-notifications/publish-events-to-amqp.rst:180]: So I wasn't sure if the majority of S3-API SDKs out there had built-in retry logic. I assumed at minimum they would throw errors or logs during downtime events, and figured it was worth mentioning in case users get surprised by a sudden burst of log messages during a restart.

If we *can* confidently state that applications transparently retry interrupted or 404'd operations, then I can tweak this slightly to add that.

**Comment by ravindk89** [comment]
> > @poornas do we doc those additional events anywhere?
> > I was able to derive the current list from [this doc](https://github.com/minio/minio/tree/master/docs/bucket/notifications) - is there a better source?
> 
> No, this is the complete list. http://127.0.0.1:8000/concepts/bucket-notifications.html does not show all the events - it seems to be the intent to leave monitoring events such as replication/ilm to the monitoring page. Just suggesting linking to it since the page showing how to configure notifications doesn't list them all. If it is still wip - please feel free to ignore.

Ahhh I understand what you're saying - yes, I can definitely do that. I'm not sold on keeping that high level `/concepts/bucket-notifications` page, vs just summarizing the feature on the 'Server Features' page and pointing users to Monitoring. Still chewing, but will link either way.

File: source/monitoring/bucket-notifications/publish-events-to-amqp.rst (hunk 3 above)
Comment (ravindk89) [reply source/monitoring/bucket-notifications/publish-events-to-amqp.rst:175]: Addressed

**Comment by ravindk89** [comment]
@eco-minio changes addressed. You'll need to refetch, rebase, and rebuild to see them. The `/includes/common-admonitions.rst` contains the new admonition changes. The `common-` prefixed files are there to help centralize single-sourced content, to make it easier in the future when we need to change things up.

@poornas if this all looks good your PR Approval would be much appreciated :)

**Review by eco-minio — APPROVED** [review]
LGTM

**Review by poornas — APPROVED** [review]
LGTM

**Comment by ravindk89** [comment]
I'm merging this into a feature branch https://github.com/minio/docs/tree/bucket-notifications - any remaining reviewers can comment here and I'll integrate that feedback into the feature branch.
